From 165fb333c9d954fec636dc0f1917ba50417478c0 Mon Sep 17 00:00:00 2001
From: Mark McLoughlin <markmc@redhat.com>
Date: Sat, 15 Aug 2009 19:38:15 +0100
Subject: [PATCH] Fix list updating after disk/network hot-unplug

The current code makes a poor effort at updating the device arrays after
hot-unplug. Fix that and combine the two code paths into one.

* src/qemu_driver.c: fix list updating in qemudDomainDetachNetDevice() and
  qemudDomainDetachPciDiskDevice()

(cherry picked from commit 4e12af5623e4a962a6bb911af06fa29aa85befba)

Fedora-patch: libvirt-fix-device-list-update-after-detach.patch
---
 src/qemu_driver.c |   38 ++++++++++++++++++--------------------
 1 files changed, 18 insertions(+), 20 deletions(-)

diff --git a/src/qemu_driver.c b/src/qemu_driver.c
index bd58435..2c4fd6f 100644
--- a/src/qemu_driver.c
+++ b/src/qemu_driver.c
@@ -5404,18 +5404,17 @@ try_command:
         goto cleanup;
     }
 
-    if (vm->def->ndisks > 1) {
-        vm->def->disks[i] = vm->def->disks[--vm->def->ndisks];
-        if (VIR_REALLOC_N(vm->def->disks, vm->def->ndisks) < 0) {
-            virReportOOMError(conn);
-            goto cleanup;
-        }
-        qsort(vm->def->disks, vm->def->ndisks, sizeof(*vm->def->disks),
-              virDomainDiskQSort);
-    } else {
-        VIR_FREE(vm->def->disks[0]);
-        vm->def->ndisks = 0;
+    if (i != --vm->def->ndisks)
+        memmove(&vm->def->disks[i],
+                &vm->def->disks[i+1],
+                sizeof(*vm->def->disks) * (vm->def->ndisks-i));
+    if (VIR_REALLOC_N(vm->def->disks, vm->def->ndisks) < 0) {
+        virReportOOMError(conn);
+        goto cleanup;
     }
+    qsort(vm->def->disks, vm->def->ndisks, sizeof(*vm->def->disks),
+          virDomainDiskQSort);
+
     ret = 0;
 
 cleanup:
@@ -5503,16 +5502,15 @@ qemudDomainDetachNetDevice(virConnectPtr conn,
 
     DEBUG("%s: host_net_remove reply: %s", vm->def->name,  reply);
 
-    if (vm->def->nnets > 1) {
-        vm->def->nets[i] = vm->def->nets[--vm->def->nnets];
-        if (VIR_REALLOC_N(vm->def->nets, vm->def->nnets) < 0) {
-            virReportOOMError(conn);
-            goto cleanup;
-        }
-    } else {
-        VIR_FREE(vm->def->nets[0]);
-        vm->def->nnets = 0;
+    if (i != --vm->def->nnets)
+        memmove(&vm->def->nets[i],
+                &vm->def->nets[i+1],
+                sizeof(*vm->def->nets) * (vm->def->nnets-i));
+    if (VIR_REALLOC_N(vm->def->nets, vm->def->nnets) < 0) {
+        virReportOOMError(conn);
+        goto cleanup;
     }
+
     ret = 0;
 
 cleanup:
-- 
1.6.2.5

